src/main/java/com/google/netpcapanalysis/dao/PCAPParserDaoImpl.java

```diff
@@ -0,0 +1,166 @@
+  private String MYIP = "";
+  private boolean first = true;
+
+  public PCAPParserDaoImpl(String file) { 
+      this.FILENAME = file;
+  }
+
+   /* Reads PCAP file (from file name) as a stream and puts unique connections into allPCAP HashMap. */
+  public void parseRaw() throws IOException {
+    final InputStream stream = PCAPParserDaoImpl.class.getClassLoader().getResourceAsStream(FILENAME);
+    final Pcap pcap = Pcap.openStream(stream);
+
+    pcap.loop(new PacketHandler() {
+      @Override
+      public boolean nextPacket(final Packet packet) throws IOException {
+      if(packet.hasProtocol(Protocol.IPv4)) {
+        IPPacket ip = (IPPacket) packet.getPacket(Protocol.IPv4);
+
+        String srcip = ip.getSourceIP();
+        String dstip = ip.getDestinationIP();
+
+        Buffer buffer = ip.getPayload();
+        int size = buffer.getReadableBytes();
+
+        String protocol = getProtocol(ip); 
+
+        if (first) {
+            MYIP = srcip;
+            first = false;
+        }
+        
+        // PCAPdata takes in (source, destination, domain, location, protocol, size, flagged, frequency) 
+        PCAPdata rawPCAP = new PCAPdata(srcip, dstip, "", "", protocol, size, false, 1);
+        allPCAP.add(rawPCAP);
+      }
+      return true;
+    }
+   });
+   pcap.close();
+  } 
+
+  /* Determine the protocol of IPv4 packet; if unknown, then returns "IPv4" as protocol.
+    * Checks for the following protocols — ICMP, IGMP, TCP, UDP, SCTP, SIP, SDP, ETHERNET_II, SLL, IPv4, RTP —
+    * which are the only protocols supported by pkts library. */
+  private String getProtocol(IPPacket packet) throws IOException {
+    String protocol = "IPv4";
+    // If packet has UDP or TCP protocol, fetch the port numbers (source/destination).
+    if (packet.hasProtocol(Protocol.UDP)) {
+      protocol = "UDP";
+      UDPPacket udpPacket = (UDPPacket) packet.getPacket(Protocol.UDP);
+      int dstport = udpPacket.getDestinationPort();
+      int srcport = udpPacket.getSourcePort();
+    }
+    else if (packet.hasProtocol(Protocol.TCP)) {
+      protocol = "TCP";
+      TCPPacket tcpPacket = (TCPPacket) packet.getPacket(Protocol.TCP);
+      int dstport = tcpPacket.getDestinationPort();
+      int srcport = tcpPacket.getSourcePort();
+    } 
+    else if (packet.hasProtocol(Protocol.ICMP)) {
+      protocol = "ICMP";
+    }
+    else if (packet.hasProtocol(Protocol.IGMP)) {
+      protocol = "IGMP";
+    }
+    else if (packet.hasProtocol(Protocol.SCTP)) {
+      protocol = "SCTP";
+    }
+    else if (packet.hasProtocol(Protocol.SIP)) {
+      protocol = "SIP";
+    }
+    else if (packet.hasProtocol(Protocol.SDP)) {
+      protocol = "SDP";
+    }
+    else if (packet.hasProtocol(Protocol.ETHERNET_II)) {
+      protocol = "ETHERNET_II";
+    }
+    else if (packet.hasProtocol(Protocol.SLL)) {
+      protocol = "SLL";
+    }
+    return protocol;
+  }
+ 
+  /* Removes duplicated data and increments corresponding frequencies */
```

Comment:
In the process of removing duplicates are you not accidentally removing other information? You are assuming that between a pair of SRC and DEST the protocol, size etc will also be same, which may not be true.

Reply:
The fundamental problem this is getting complicated is because you are tying frequency to each PCAP data row, whereas if you had just the PCAP data as it is, then you can store the frequencies separately in a entity something like: `srcip, destip, frequency`. The ip-pair could be be directional and if you want to aggregate them into undirectional that should be fairly easy if you loop through each unique `srcip` in the list and check if it is in a `destip`, then increment the map where that row's srcip is in current rows destip.